packages/appsync-modelgen-plugin/src/utils/process-connections-v2.ts

```diff
@@ -0,0 +1,103 @@
+
+export function getConnectedFieldV2(field: CodeGenField, model: CodeGenModel, connectedModel: CodeGenModel, directiveName: string): CodeGenField {
+  const connectionInfo = getDirective(field)(directiveName);
+  if (!connectionInfo) {
+    throw new Error(`The ${field.name} on model ${model.name} is not connected`);
+  }
+
+  const connectionName = connectionInfo.arguments.name;
+  const indexName = connectionInfo.arguments.indexName;
+  const connectionFields = connectionInfo.arguments.fields;
+  if (connectionFields) {
+    let keyDirective;
+    if (indexName) {
+      keyDirective = flattenFieldDirectives(connectedModel).find(dir => {
+        return dir.name === 'index' && dir.arguments.name === indexName;
+      });
+      if (!keyDirective) {
+        throw new Error(
+          `Error processing @${connectionInfo.name} directive on ${model.name}.${field.name}, @index directive with name ${indexName} was not found in connected model ${connectedModel.name}`,
+        );
+      }
+    } else {
+      keyDirective = flattenFieldDirectives(connectedModel).find(dir => {
+        return dir.name === 'primaryKey';
+      });
+    }
+
+    // when there is a fields argument in the connection
+    const connectedFieldName = keyDirective ? ((fieldDir: CodeGenFieldDirective) => { return fieldDir.fieldName ;})(keyDirective as CodeGenFieldDirective) : DEFAULT_HASH_KEY_FIELD;
+
+    // Find a field on the other side which connected by a @connection and has the same fields[0] as indexName field
+    const otherSideConnectedField = connectedModel.fields.find(f => {
+      return f.directives.find(d => {
+        return (d.name === 'belongsTo' || d.name === 'hasOne' || d.name === 'hasMany') && d.arguments.fields && d.arguments.fields[0] === connectedFieldName;
+      });
+    });
+    if (otherSideConnectedField) {
+      return otherSideConnectedField;
+    }
+    // If there are no field with @connection with indexName then try to find a field that has same name as connection name
+    const connectedField = connectedModel.fields.find(f => f.name === connectedFieldName);
+
+    if (!connectedField) {
+      throw new Error(`Can not find key field ${connectedFieldName} in ${connectedModel}`);
+    }
+    return connectedField;
+  } else if (connectionName) {
+    // when the connection is named
+    const connectedField = connectedModel.fields.find(f =>
+      f.directives.find(d => (d.name === 'belongsTo' || d.name === 'hasOne' || d.name === 'hasMany') && d.arguments.name === connectionName && f !== field),
+    );
+    if (!connectedField) {
+      throw new Error(`Can not find key field with connection name ${connectionName} in ${connectedModel}`);
+    }
+    return connectedField;
+  }
+  // un-named connection. Use an existing field or generate a new field
+  const connectedFieldName = makeConnectionAttributeName(model.name, field.name);
```

Comment:
I am not sure about this behavior when we cannot find the connected field. How does the new field that is generated gets populated?

Reply:
I think it's generated by the transformer in the dynamo tables, as far as the codegen work goes I was only trying to replicate the existing behavior with new directives.

Reply:
Can you please include a unit test for this if one is not already there.

Reply:
We have E2E tests covering this in the cli since it's something that plays off the amplify-cli repo, unit testing just in the scope of codegen doesn't cover enough for this case